text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,482 @@
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  P[Publish]-->CM1[Create m1]
+  P-->CM2[Create m2]
+  end
+  subgraph CONSUMER
+  direction TB
+  RM1[Receive m1]
+  RM2[Receive m1]
+  end
+  CM1-. link .->RM1;
+  CM2-. link .->RM2;
+```
+
+### Batch message producer, single message push-based consumer
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  P[Publish]
+  end
+  subgraph CONSUMER
+  direction TB
+  DM1[Deliver m1]
+  DM2[Deliver m1]
+  end
+  P-. link .->DM1;
+  P-. link .->DM2;
+```
+
+### Batch message producer with "Create" spans populated before publish, single message pull-based consumer
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  A[Ambient]-->CM1[Create m1]
+  A-->CM2[Create m2]
+  A-->P[Publish]
+  end
+  subgraph CONSUMER
+  direction TB
+  RM1[Receive m1]
+  RM2[Receive m1]
+  end
+  CM1-. link .->RM1;
+  CM2-. link .->RM2;
+```
+
+### Single message producers, batch push-based consumer with process spans
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  PM1[Publish m1]
+  PM2[Publish m2]
+  end
+  subgraph CONSUMER
+  direction TB
+  D[Deliver]-->PRM1[Process m1]
+  D-->PRM2[Process m2]
+  end
+  PM1-. link .->D;
+  PM2-. link .->D;
+```
+
+### Single message producers, batch pull-based consumer with process spans
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  PM1[Publish m1]
+  PM2[Publish m2]
+  end
+  subgraph CONSUMER
+  direction TB
+  A[Ambient]-->R[Receive]
+  A-->PRM1[Process m1]
+  A-->PRM2[Process m2]
+  end
+  PM1-. link .->R;
+  PM2-. link .->R;
+```
+
+### Single message producers, batch pull-based consumer with manual settlement
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  PM1[Publish m1]
+  PM2[Publish m2]
+  end
+  subgraph CONSUMER
+  direction TB
+  A[Ambient]-->R[Receive]
+  A-->SM1[Settle m1]
+  A-->SM2[Settle m2]
+  end
+  PM1-. link .->R;
+  PM2-. link .->R;
+```
+
+## Future possibilities
+
+### Intermediary instrumentation
+
+While intermediary instrumentation is not directly covered by the conventions
+in this document, it certainly is necessary to keep the proposed conventions
+extensible so that intermediary instrumentation can be easily added in a way
+that integrates well with producer and consumer instrumentation.
+
+The diagram below gives an example of how intermediary instrumentation can be
+added. The fact that producers and consumers are consistently correlated by
+links across all scenarios provides maximal flexibility for adding intermediary
+instrumentation.
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  PM1[Publish m1]
+  PM2[Publish m2]
+  end
+  subgraph INTERMEDIARY
+  direction TB
+  EQ[Enqueue]
+  DQ[Dequeue]
+  end
+  subgraph CONSUMER
+  direction TB
+  D[Deliver]-->PRM1[Process m1]
+  D-->PRM2[Process m2]
+  end
+  PM1-. link .->D;
+  PM2-. link .->D;
+  PM1-->EQ;
+  PM2-->EQ;
+  DQ-->D;
```

Comment:
Not sure this is a clear picture explaining how things work at messaging level. It depends on what we mean as "enqueue" and "dequeue" because when you have an intermediary in the middle, in the end you have a contract between producer and intermediary, and one between intermediary and consumer. 
So the intermediary "receive"(s) the message from producer (so you could see it as a consumer) and then it "publish"(es) the message to the consumer (so you could see it as a producer).

Reply:
I can change those names. I didn't want to use "Publish" and "Receive", as those are names we use for producer and consumer spans, so I wanted to avoid confusion.

Maybe I should just have empty boxes in the intermediary to avoid any confusion?

Reply:
I would just remove those boxes and not having them empty. Just having a "gray" intermediary box without the need to highlight what's in there (even because tracing within intermediary is out of scope for now).

Reply:
I'd like to leave the example here, because it shows that the proposed the solution can easily extended to accomodate intermediary instrumentation.

The boxes inside the intermediary just mean to highlight that there exist some (yet undefined) span in the intermediary that acts as parent/child for the consumer and producer spans.

However, I'm open to further adjust the diagram if it makes things clearer.

Reply:
> I'd like to leave the example here, because it shows that the proposed the solution can easily extended to accomodate intermediary instrumentation.

I am totally fine with leaving the example.

> The boxes inside the intermediary just mean to highlight that there exist some (yet undefined) span in the intermediary that acts as parent/child for the consumer and producer spans.

Tbh the empty boxes can just confuse people because they don't add/clarify nothing.